### ISSUE TITLE
False positive of incompatible type report with StrEnum and str

### ISSUE THREAD

**Issue by ksamuel** [python/mypy#14230]
**Bug Report**

When using StrEnum as a type but the litteral str as a value, mypy reports it as the wrong type.

**To Reproduce**

Gist: https://gist.github.com/mypy-play/c24dd4a6c2b9883a27dfd60a63bbc19a

```python
from enum import StrEnum, auto

class Foo(StrEnum):
	bar = auto()

def f(param: Foo):
	pass

f("bar")
```

**Expected Behavior**

mypy should check that "bar" is a member of the StrEnum, and report no error, as it would do if `Literal["bar"]` were used in place of StrEnum.

**Actual Behavior**

```
script.py:11: error: Argument 1 to "f" has incompatible type "str"; expected "Foo"  [arg-type]
```

**Your Environment**

python 3.11 with mypy 0.991

**Comment by JelleZijlstra** [comment]
Mypy is correct here. `"bar"` is an instance of `str`, not `Foo`. `Foo` is a subclass of `str`, but that doesn't change the type of existing strings.

**Comment by ksamuel** [comment]
The message is correct, but the behavior is not duck typing friendly.

Just like we have structural sub-typing to accept objects with equivalent behavior but different type, I was surprised by mypy rejected this code as unsafe.

Indeed:

- `assert Foo.bar == "bar"` is True
- "bar" is a member of Foo, so completeness is respected
- match/case would match Foo.bar and "bar" without distinction

For all practical use cases of StrEnum, a code accepting either value in the spirit of duck typing would run sanely. 

Do you think it's not legitimate that mypy considers this code as type safe ?

**Comment by erictraut** [comment]
No, this isn't type safe. Here's a trivial example that crashes at runtime.

```python
def f(param: Foo):
	print(param.value)

f("bar")
```

**Comment by AlexWaygood** [comment]
@ksamuel, you have the subtyping relationship backwards. `Foo` is both a subclass and a subtype of `str`, so mypy will happily accept members of `Foo` for a parameter that's annotated as accepting `str`. But `str` is neither a subclass nor a subtype of `Foo`, as @erictraut  shows.

**Comment by ksamuel** [comment]
I do understand this, but duck typing is not related to inheritance, only the common behavior you want to accept. 

However, I can see that I am reporting something as a bug that is working as intended.

I just have no object in Python allowing me to do this unless I provide a union with a Literal.

**Comment by takeda** [comment]
@ksamuel by default mypy uses nominal typing, you can use structural by using protocols: https://mypy.readthedocs.io/en/stable/protocols.html

**Comment by ksamuel** [comment]
I indeed mention protocols in previous comments, but they cannot be used in this case, since protocols can check something is string like, but cannot check the value is part of the enum. Now again as mentioned, the only way to allow this is to use a Union of Literal and StrEnum, but then it gets incredibly verbose and you risk a bug if you remove a value from the enum but not the literal by mistake.

**Comment by erictraut** [comment]
If you want to use a literal string, I recommend using a union of literal strings for typing purposes. If you want to use a `StrEnum`, then you should use the enum values from the enum (e.g. `f(Foo.bar)`). If you use one approach or the other, it will work well with static typing. The problem is that you're trying to combine the two approaches, and that's not going to work.

Approach 1 with String Literals:
```python
from typing_extensions import Literal

Foo = Literal['foo', 'bar']

def f(param: Foo):
	pass

f("bar")
```

Approach 2 with StrEnum:
```python
from enum import StrEnum, auto

class Foo(StrEnum):
	foo = auto()
	bar = auto()

def f(param: Foo):
	pass

f(Foo.bar)
```
